src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -32,48 +34,50 @@ const vscodeNotebookEnums = require('vscode') as typeof import('vscode-proposed'
+    private readonly documentExecutions = new WeakMap<NotebookDocument, CancellationTokenSource>();
+    private readonly stackOfCellsToExecuteByDocument = new WeakMap<NotebookDocument, CellExecution[]>();
 
     private readonly kernelValidated = new WeakMap<NotebookDocument, { kernel: IKernel; promise: Promise<void> }>();
 
     private readonly executionFactory: CellExecutionFactory;
     private readonly disposables: IDisposable[] = [];
+    private readonly pendingExecution = new ChainedExecutions<void>('Kernel execution failure');
+    private readonly pendingCellExecution = new ChainedExecutions<NotebookCellRunState | undefined>(
+        'Kernel execution previous cell failure'
+    );
     private isRawNotebookSupported?: Promise<boolean>;
+    private _interruptPromise?: Promise<InterruptResult>;
     constructor(
         private readonly kernelProvider: IKernelProvider,
-        private readonly commandManager: ICommandManager,
         errorHandler: IDataScienceErrorHandler,
         editorProvider: INotebookEditorProvider,
         readonly kernelSelectionUsage: IKernelSelectionUsage,
         readonly appShell: IApplicationShell,
         readonly vscNotebook: IVSCodeNotebook,
         readonly metadata: Readonly<KernelConnectionMetadata>,
         private readonly rawNotebookSupported: IRawNotebookSupportedService,
-        context: IExtensionContext
+        context: IExtensionContext,
+        private readonly interruptTimeout: number
     ) {
         this.executionFactory = new CellExecutionFactory(errorHandler, editorProvider, appShell, vscNotebook, context);
     }
 
     @captureTelemetry(Telemetry.ExecuteNativeCell, undefined, true)
     public async executeCell(notebookPromise: Promise<INotebook>, cell: NotebookCell): Promise<void> {
-        // Cannot execute empty cells.
-        if (this.cellExecutions.has(cell) || cell.document.getText().trim().length === 0) {
-            // clear cell output
-            cell.outputs = [];
+        // Return current execution.
+        if (this.cellExecutions.get(cell)) {
+            await this.cellExecutions.get(cell)!.result;
             return;
         }
         const editor = this.vscNotebook.notebookEditors.find((item) => item.document === cell.notebook);
         if (!editor) {
             return;
         }
-        const cellExecution = this.executionFactory.create(cell, isPythonKernelConnection(this.metadata));
-        this.cellExecutions.set(cell, cellExecution);
-
+        const cellExecution = this.createCellExecution(cell);
         try {
-            const notebook = await notebookPromise;
-            const kernel = this.getKernel(cell.notebook);
-            traceCellMessage(cell, 'executeCell started in KernelExecution');
-            await this.executeIndividualCell(kernel, cellExecution, notebook);
+            await this.pendingExecution.chainExecution(async () =>
+                this.executeQueuedCells(notebookPromise, cell.notebook)
```

Comment:
Fixes to how we queued cells.
Whether you run the whole notebook or individual notebook, they all go into the same queue.
problem - run all cells & then run an individual cell was causing problems, this ensures they are queued & things don't go out of order.